adapters/condition_pthreads.c

```diff
@@ -1,150 +1,157 @@
-            // handle overflow in tv_nsec
-            tm.tv_sec+= tm.tv_nsec / NANOSECONDS_IN_1_SECOND;
-            tm.tv_nsec %= NANOSECONDS_IN_1_SECOND;
-            int wait_result = pthread_cond_timedwait((pthread_cond_t *)handle, (pthread_mutex_t *)lock, &tm);
-            if (wait_result == ETIMEDOUT)
-            {
-                // Codes_SRS_CONDITION_18_011: [ Condition_Wait shall return COND_TIMEOUT if the condition is NOT triggered and timeout_milliseconds is not 0 ]
-                return COND_TIMEOUT;
-            }
-            else if (wait_result == 0)
-            {
-                // Codes_SRS_CONDITION_18_012: [ Condition_Wait shall return COND_OK if the condition is triggered and timeout_milliseconds is not 0 ]
-                result = COND_OK;
-            }
-            else
-            {
-                LogError("Failed to pthread_cond_timedwait");
-                return COND_ERROR;
-            }
-        }
-        else
-        {
-            if (pthread_cond_wait((pthread_cond_t*)handle, (pthread_mutex_t *)lock) != 0)
-            {
-                LogError("Failed to pthread_cond_wait");
-                return COND_ERROR;
-            }
-            else
-            {
-                // Codes_SRS_CONDITION_18_010: [ Condition_Wait shall return COND_OK if the condition is triggered and timeout_milliseconds is 0 ]
-                result = COND_OK;
-            }
-        }
-    }
-    return result;
-}
-
-void Condition_Deinit(COND_HANDLE handle)
-{
-// Codes_SRS_CONDITION_18_007: [ Condition_Deinit will not fail if handle is NULL ]
-    if (handle != NULL)
-    {
-        // Codes_SRS_CONDITION_18_009: [ Condition_Deinit will deallocate handle if it is not NULL 
-        pthread_cond_t* cond = (pthread_cond_t*)handle;
-        pthread_cond_destroy(cond);
-        free(cond);
-    }
-}
-
-
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include "azure_c_shared_utility/lock.h"
+#include "azure_c_shared_utility/condition.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include <errno.h>
+#include <pthread.h>
+#include "azure_c_shared_utility/gballoc.h"
+#include "pthread.h"
+#include "time.h"
+#include "mach_gettime.h"
+
+DEFINE_ENUM_STRINGS(COND_RESULT, COND_RESULT_VALUES);
+
+// Set our time basis based on what clock is available.  Give preference to CLOCK_MONOTONIC,
+// then CLOCK_REALTIME.  If neither is available, query the default pthread_condattr_t value 
+// and use that.
+#ifdef CLOCK_MONOTONIC 
+clockid_t time_basis = CLOCK_MONOTONIC;
+#else 
+#ifdef CLOCK_REALTIME
+clockid_t time_basis = CLOCK_REALTIME;
+#else
+clockid_t time_basis = -1;
+#endif
+#endif
+
+COND_HANDLE Condition_Init(void)
+{
+#if TARGET_OS_IPHONE
+    //pthread_condattr_getclock and pthread_condattr_setclock not available on iOS 9.3
+    return NULL;
+#else
```

Comment:
The condition code should be turned off by default.
In CMakeLists:

option(use_condition "set use_condition to ON if the condition module and its adapters should be enabled" OFF)

It is used in the IoT Gateway project only, so unless you would like to port that the changes in this file and the addition of mach_gettime should not be necessary.